### PR TITLE
close #177 assign `float3_X = float_X`

### LINKED ISSUE #177: Assign float3_X = float_X should not be allowed
As seen [here](https://github.com/psychocoderHPC/picongpu/commit/ebe3460589db1af1721c813de6a8114abc76e882) small errors can have big effects.

Right now *,
```c++
float_X b( 1.0 );
float3_X a = b;
```
does *not* throw an error. That is super-risky and should **not be allowed**.
(Same for all other assign related operators like `+=`, `-=`, `*=`, `/=`, ...)

(*) now means state of commit 6e5905f1c029a4c7ef12eca0057b665705f80602

### PR DESCRIPTION
**Pull request by psychocoderHPC**
close #177 **Assign `float3_X = float_X` should not be allowed**

- add method to create vector were all components were set
to the same value `math::vector<...>`
- remove ambiguous `vector<T_Type,...>(T_Type value)` constructor
- add new constructor for one dimensional vectors
- fix dependencies

**Do not merge before #807**

### CODE REVIEW

File: src/libPMacc/include/cuSTL/container/CartBuffer.tpp [hunk 1]
```diff
@@ -199,7 +199,7 @@ CartBuffer<Type, T_dim, Allocator, Copier, Assigner>::view
 {
     a = (a + (math::Int<T_dim>)this->size()) % (math::Int<T_dim>)this->size();
     b = (b + (math::Int<T_dim>)this->size())
-            % ((math::Int<T_dim>)this->size() + math::Int<T_dim>(1));
+            % ((math::Int<T_dim>)this->size() + math::Int<T_dim>(math::Int<T_dim>::create(1)));
```
Comment (psychocoderHPC) [inline src/libPMacc/include/cuSTL/container/CartBuffer.tpp]: not needed double cast (thx @BenjaminW3 )

File: src/libPMacc/include/cuSTL/cursor/MultiIndexCursor.hpp [hunk 2]
```diff
@@ -43,7 +43,7 @@ template<int dim>
 HDINLINE
 cursor::Cursor<cursor::MarkerAccessor<math::Int<dim> >, MultiIndexNavigator<dim>,
                math::Int<dim> >
-               make_MultiIndexCursor(const math::Int<dim>& idx = math::Int<dim>(0))
+               make_MultiIndexCursor(const math::Int<dim>& idx = math::Int<dim>(math::Int<dim>::create(0)))
```
Comment (BenjaminW3) [inline src/libPMacc/include/cuSTL/cursor/MultiIndexCursor.hpp]: double cast

File: src/libPMacc/include/cuSTL/zone/SphericZone.hpp [hunk 3]
```diff
@@ -53,7 +53,7 @@ struct SphericZone
     math::Int<dim> offset;
 
     HDINLINE SphericZone() {}
-    HDINLINE SphericZone(const math::Size_t<dim>& size) : size(size), offset(math::Int<dim>(0)) {}
+    HDINLINE SphericZone(const math::Size_t<dim>& size) : size(size), offset(math::Int<dim>(math::Int<dim>::create(0))) {}
```
Comment (BenjaminW3) [inline src/libPMacc/include/cuSTL/zone/SphericZone.hpp]: double cast

File: src/picongpu/include/fields/laserProfiles/laserGaussianBeam.hpp [hunk 4]
```diff
@@ -43,7 +43,7 @@ namespace picongpu
             const double runTime = DELTA_T*currentStep;
             const double f = SPEED_OF_LIGHT / WAVE_LENGTH;
 
-            float3_X elong = float3_X(float_X(0.0), float_X(0.0), float_X(0.0));
+            float3_X elong = float3_X(float3_X::create(0.0));
```
Comment (BenjaminW3) [inline src/picongpu/include/fields/laserProfiles/laserGaussianBeam.hpp]: double cast

File: src/picongpu/include/fields/laserProfiles/laserNone.hpp [hunk 5]
```diff
@@ -39,7 +39,7 @@ namespace picongpu
          */
         HINLINE float3_X laserLongitudinal( uint32_t, float_X& phase )
         {
-            float3_X elong = float3_X(float_X(0.0), float_X(0.0), float_X(0.0));
+            float3_X elong = float3_X(float3_X::create(0.0));
```
Comment (BenjaminW3) [inline src/picongpu/include/fields/laserProfiles/laserNone.hpp]: double cast

File: src/picongpu/include/fields/laserProfiles/laserPlaneWave.hpp [hunk 6]
```diff
@@ -44,7 +44,7 @@ namespace picongpu
             const double f = SPEED_OF_LIGHT / WAVE_LENGTH;
 
             double envelope = double(AMPLITUDE );
-            float3_X elong = float3_X(float_X(0.0), float_X(0.0), float_X(0.0));
+            float3_X elong = float3_X(float3_X::create(0.0));
```
Comment (BenjaminW3) [inline src/picongpu/include/fields/laserProfiles/laserPlaneWave.hpp]: double cast

File: src/picongpu/include/fields/laserProfiles/laserPulseFrontTilt.hpp [hunk 7]
```diff
@@ -43,7 +43,7 @@ namespace picongpu
             const double runTime = DELTA_T*currentStep;
             const double f = SPEED_OF_LIGHT / WAVE_LENGTH;
 
-            float3_X elong = float3_X(float_X(0.0), float_X(0.0), float_X(0.0));
+            float3_X elong = float3_X(float3_X::create(0.0));
```
Comment (BenjaminW3) [inline src/picongpu/include/fields/laserProfiles/laserPulseFrontTilt.hpp]: double cast

File: src/picongpu/include/fields/laserProfiles/laserWavepacket.hpp [hunk 8]
```diff
@@ -40,7 +40,7 @@ namespace laserWavepacket
 HINLINE float3_X laserLongitudinal(uint32_t currentStep, float_X& phase)
 {
     float_X envelope = float_X(AMPLITUDE);
-    float3_X elong = float3_X(float_X(0.0), float_X(0.0), float_X(0.0));
+    float3_X elong = float3_X(float3_X::create(0.0));
```
Comment (BenjaminW3) [inline src/picongpu/include/fields/laserProfiles/laserWavepacket.hpp]: double cast

**Comment by psychocoderHPC** [comment]
fix used double casts

**Comment by psychocoderHPC** [comment]
wait for #807

**Comment by PrometheusPi** [comment]
#807 was merged :smile:

**Comment by ax3l** [comment]
@BenjaminW3 usually, you should wait until @psychocoderHPC rebases against `dev` so the commit a736f15002fff3a4bafc9126c582c94fc97416b0 does not appear any more (that is not a git problem but a readability-thing in github).

the merge was a tiny bit too early ;) next time.

**Pull request merged**